### PR TITLE
Update Dataclasses docs

### PR DESCRIPTION
**Pull request by tpdorsey**
- document new dataclasses features
- review docstrings
- update typeadapter links

## Related issue number

Closes PYD-54

## Checklist

* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [ ] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @davidhewitt

### CODE REVIEW

**Comment by tpdorsey** [comment]
please review

File: docs/migration.md [hunk 1]
```diff
@@ -145,23 +145,27 @@ The following properties have been removed from or changed in `Field`:
 
 ### Changes to dataclasses
 
+Pydantic [dataclasses](usage/dataclasses.md) continue to be useful for enabling the data validation on standard
+dataclasses without having to subclass `BaseModel`. However, we have made some changes to the behavior of Pydantic
+dataclasses in V2:
```
Comment (ybressler) [inline docs/migration.md:150]: I haven't seen `we` used in the docs yet – perhaps change language to be more subjective.

```suggestion
Pydantic [dataclasses](usage/dataclasses.md) continue to be useful for enabling the data validation on standard
dataclasses without having to subclass `BaseModel`. Pydantic V2 introduces the following changes to this dataclass behavior:
```

Comment (ybressler) [reply docs/migration.md:150]: Oh, lol, I do see `we` used... Welp.

File: docs/migration.md [hunk 2]
```diff
@@ -145,23 +145,27 @@ The following properties have been removed from or changed in `Field`:
...
+
 * When used as fields, dataclasses (Pydantic or vanilla) no longer accept tuples as validation inputs; dicts should be
   used instead.
 * The `__post_init__` in Pydantic dataclasses will now be called _after_ validation, rather than before.
     * As a result, the `__post_init_post_parse__` method would have become redundant, so has been removed.
 * We no longer support `extra='allow'` for Pydantic dataclasses, where extra fields passed to the initializer would be
```
Comment (ybressler) [inline docs/migration.md:156]: Nit: remove `we` (I see it exists before your edits):
```
* Pydantic dataclasses longer support for `extra='allow'` where extra fields passed to the initializer would be
```

File: docs/migration.md [hunk 3]
```diff
@@ -145,23 +145,27 @@ The following properties have been removed from or changed in `Field`:
...
+        any other functionality that may have required `__pydantic_model__` in V1, you should now wrap the dataclass
+        with a [`TypeAdapter`](usage/models.md#typeadapter) ([discussed more below](#introduction-of-typeadapter)) and
+        make use of its methods.
 * In Pydantic V1, if you used a vanilla (i.e., non-Pydantic) dataclass as a field, the config of the parent type would
-  be used as though it was the config for the dataclass itself as well. In Pydantic V2, this is no longer the case.
-    * [TODO: Need to specify how to override the config used for vanilla dataclass; possibly need to add functionality?]
-
+    be used as though it was the config for the dataclass itself as well. In Pydantic V2, this is no longer the case.
+    * In Pydantic V2, to override the config like you would with a `BaseModel`, you can use the `config` parameter
+        on the `dataclass` decorator. See [Dataclass Config](usage/dataclasses.md#dataclass-config) for examples.
```
Comment (ybressler) [inline docs/migration.md:168]: _(This one is personal style related - feel free to ignore)_
```
    * In Pydantic V2, to override the `config` (like you would with a `BaseModel`), you can use the `config` parameter
        on the `dataclass` decorator. See [Dataclass Config](usage/dataclasses.md#dataclass-config) for examples.
```

Comment (ybressler) [inline docs/migration.md:168]: Is it `config` or `model_config`?

File: docs/usage/dataclasses.md [hunk 4]
```diff
@@ -25,18 +25,26 @@ User(id=42, name='John Doe', signup_ts=datetime.datetime(2032, 6, 21, 12, 0))
...
+*  How [initialization hooks](#initialization-hooks) work
+*  [JSON dumping](#json-dumping)
+
+You can use all the standard Pydantic field types. Note, however, that arguments passed to constructor will be copied in
+order to perform validation and, where necessary coercion.
+
+To perform validation or generate a JSON schema on a Pydantic dataclass, you should now wrap the dataclass
+with a [`TypeAdapter`](models.md#typeadapter) and make use of its methods.
+
+Fields that require a `default_factory` can be specified by either a `pydantic.Field` or a `dataclasses.field`.
```
Comment (ybressler) [inline docs/usage/dataclasses.md:47]: Should these be bullet points? (Not a suggestion, a Q)

File: docs/usage/dataclasses.md [hunk 5]
```diff
@@ -260,8 +272,8 @@ except dataclasses.FrozenInstanceError as e:
 
 ### Use custom types
 
-Since stdlib dataclasses are automatically converted to add validation using
-custom types may cause some unexpected behaviour.
+Since stdlib dataclasses are automatically converted to add validation, using
+custom types may cause some unexpected behavior.
```
Comment (ybressler) [inline docs/usage/dataclasses.md:276]: Small diff between US/UK english syntax. I don't think it matters – but we should probably be consistent.

File: docs/usage/dataclasses.md [hunk 6]
```diff
@@ -317,10 +329,10 @@ print(repr(m))
 #> Model(dc=DC(a=ArbitraryType(value=3), b='qwe'), other='other')
 ```
 
-## Initialize hooks
+## Initialization hooks
```
Comment (ybressler) [inline docs/usage/dataclasses.md:332]: Are these `hooks` which are initialized? _If yes_, then `Initialize hooks` is the correct title.

Are these `hooks which run on class initialization`? _If yes_, `Initialization hooks` is the correct title.

File: pydantic/dataclasses.py [hunk 7]
```diff
@@ -107,7 +107,7 @@ def dataclass(
     kw_only: bool = False,
     slots: bool = False,
 ) -> Callable[[type[_T]], type[PydanticDataclass]] | type[PydanticDataclass]:
-    """A decorator used to create a Pydantic-enhanced dataclass, similar to the standard Python `dataclasses`,
+    """A decorator used to create a Pydantic-enhanced dataclass, similar to standard Python `dataclasses`,
```
Comment (ybressler) [inline pydantic/dataclasses.py:110]: I believe `the` is more correct syntax here.

**Review by ybressler — CHANGES_REQUESTED** [review]
Thank you for going through the documentation! Your changes are scrupulous.

I left some comments + requested changes related to style + grammar. Overall, looks really good.

File: docs/migration.md (hunk 2 above)
Comment (tpdorsey) [inline docs/migration.md:156]:
```suggestion
* Pydantic no longer supports `extra='allow'` for Pydantic dataclasses, where extra fields passed to the initializer would be
```

File: docs/migration.md (hunk 3 above)
Comment (tpdorsey) [reply docs/migration.md:168]: I believe it's the `config` option on `@dataclass` in this context.

File: docs/usage/dataclasses.md (hunk 4 above)
Comment (tpdorsey) [reply docs/usage/dataclasses.md:47]: Not in this context

File: docs/usage/dataclasses.md (hunk 6 above)
Comment (tpdorsey) [reply docs/usage/dataclasses.md:332]: The latter in this case

File: pydantic/dataclasses.py (hunk 7 above)
Comment (tpdorsey) [inline pydantic/dataclasses.py:110]:
```suggestion
    """A decorator used to create a Pydantic-enhanced dataclass, similar to the standard Python `dataclasses`,
```

File: docs/migration.md [hunk 8]
```diff
@@ -145,23 +145,26 @@ The following properties have been removed from or changed in `Field`:
...
-    * [TODO: Add link to TypeAdapter documentation. You can find example usage in `tests/test_type_adapter.py`.]
+        any other functionality that may have required `__pydantic_model__` in V1, you should now wrap the dataclass
+        with a [`TypeAdapter`](usage/models.md#typeadapter) ([discussed more below](#introduction-of-typeadapter)) and
+        make use of its methods.
 * In Pydantic V1, if you used a vanilla (i.e., non-Pydantic) dataclass as a field, the config of the parent type would
-  be used as though it was the config for the dataclass itself as well. In Pydantic V2, this is no longer the case.
-    * [TODO: Need to specify how to override the config used for vanilla dataclass; possibly need to add functionality?]
-
+    be used as though it was the config for the dataclass itself as well. In Pydantic V2, this is no longer the case.
+    * In Pydantic V2, to override the config like you would with a `BaseModel`, you can use the `config` parameter
```
Comment (tpdorsey) [inline docs/migration.md:166]:
```suggestion
    * In Pydantic V2, to override the `config` (like you would with a `BaseModel`), you can use the `config` parameter
```

**Comment by tpdorsey** [comment]
please review

File: docs/migration.md (hunk 3 above)
Comment (ybressler) [reply docs/migration.md]: Right you are. Worth mentioning that this is different from BaseModels? (Or is that info cramming?)

File: pydantic/dataclasses.py [hunk 9]
```diff
@@ -107,7 +107,7 @@ def dataclass(
     kw_only: bool = False,
     slots: bool = False,
 ) -> Callable[[type[_T]], type[PydanticDataclass]] | type[PydanticDataclass]:
-    """A decorator used to create a Pydantic-enhanced dataclass, similar to standard Python `dataclasses`,
+    """A decorator used to create a Pydantic-enhanced dataclass, similar to the standard Python `dataclasses`,
```
Comment (ybressler) [inline pydantic/dataclasses.py]: _(Sorry for the back and forth here!)_ I noticed `dataclasses` is plural, which would negative `the` – here's the correct grammar for each:
* **Plural:** similar to standard Python `dataclasses`, ...
* **Singular:** similar to the standard Python `dataclass`, ...

**Review by ybressler — CHANGES_REQUESTED** [review]
Looks good! I have one small change about one word ;)

Also, one outstanding thread about how to resolve potential confusion with `model_config` for basemodels and `config` for dataclasses. Pending resolution, we should be good to go!

File: docs/migration.md (hunk 3 above)
Comment (tpdorsey) [reply docs/migration.md]: @ybressler reworded based on your feedback, hopefully this is more clear about BaseModel vs dataclass config

File: pydantic/dataclasses.py [hunk 10]
```diff
@@ -107,31 +107,31 @@ def dataclass(
...
-        order: Determines if comparison magic methods should be generated, such as` __lt__`, but not `__eq__`.
+        order: Determines if comparison magic methods should be generated, such as `__lt__`, but not `__eq__`.
         unsafe_hash: Determines if an unsafe hashing function should be included in the class.
-        frozen: Determines if the generated class should be a 'frozen' dataclass, which does not allow its
+        frozen: Determines if the generated class should be a 'frozen' `dataclass`, which does not allow its
             attributes to be modified from its constructor.
-        config: A configuration for the dataclass generation.
-        validate_on_init: A deprecated parameter included for backwards compatibility; in V2, all pydantic dataclasses
+        config: A configuration for the `dataclass` generation.
+        validate_on_init: A deprecated parameter included for backwards compatibility; in V2, all Pydantic `dataclass`es
```
Comment (ybressler) [inline pydantic/dataclasses.py:127]: Plural is okay here ->
```suggestion
validate_on_init: A deprecated parameter included for backwards compatibility; in V2, all Pydantic `dataclasses`
```

**Review by ybressler — APPROVED** [review]
Changes look great!

**Comment by tpdorsey** [comment]
Please review

**Review by ybressler — APPROVED** [review]

File: docs/usage/dataclasses.md [hunk 11]
```diff
@@ -126,11 +134,12 @@ class MyDataclass2:
...
 1. You can read more about `validate_assignment` in [model_config](model_config.md#validate-assignment).
 
-!!! warning
-    After v1.10, _pydantic_ dataclasses support `Config.extra` but some default behaviour of stdlib dataclasses
-    may prevail. For example, when `print`ing a _pydantic_ dataclass with allowed extra fields, it will still
-    use the `__str__` method of stdlib dataclass and show only the required fields.
-    This may be improved further in the future.
+!!! note
+    Pydantic dataclasses do not support [`extra='allow'`](model_config/#extra-attributes), where extra fields passed
+    to the initializer would be stored as extra attributes on the dataclass.
```
Comment (gordonhart) [inline docs/usage/dataclasses.md:139]: Hi @tpdorsey! Could you help me understand what you meant by this diff? It looks to me like the current support for `extra="allow"` is the same as it was after #2557 (the PR where the removed copy was introduced), but I'm wondering if I'm missing something.

With `extra="allow"`, extra values passed to `__init__` still end up in the instantiated object, but just aren't surfaced via `__str__`. Example script:

```python
from pydantic.dataclasses import dataclass

@dataclass(config=dict(extra="allow"))
class Example:
    ...

obj = Example(a=1, b=2)
print(obj)
print(f"Extra: a={obj.a}, b={obj.b}")
```

This script yields the following output when running with the latest `pydantic==2.3.0`:

```
Example()
Extra: a=1, b=2
```

Comment (gordonhart) [reply docs/usage/dataclasses.md:139]: Thanks for the quick response, @hramezani — opened https://github.com/pydantic/pydantic/discussions/7362 to discuss.

**Pull request merged**